File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/history/TransactionsHistory.java
package org.mifos.mobilewallet.mifospay.history;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.base.UseCaseFactory;
import org.mifos.mobilewallet.core.base.UseCaseHandler;
import org.mifos.mobilewallet.core.data.fineractcn.entity.journal.JournalEntry;
import org.mifos.mobilewallet.core.domain.usecase.journal.FetchJournalEntries;
import org.mifos.mobilewallet.mifospay.utils.Constants;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

public class TransactionsHistory {

    private final UseCaseHandler mUsecaseHandler;
    public HistoryContract.TransactionsHistoryAsync delegate;
    @Inject
    FetchJournalEntries fetchJournalEntriesUseCase;
    @Inject
    UseCaseFactory mUseCaseFactory;
    private List<JournalEntry> transactions;

    @Inject
    public TransactionsHistory(UseCaseHandler useCaseHandler) {
        this.mUsecaseHandler = useCaseHandler;
        transactions = new ArrayList<>();
    }

    public void fetchTransactionsHistory(String accountIdentifier) {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
        String currentDate = sdf.format(new Date());
        String dateRange = Constants.STARTING_DATE + ".." + currentDate + "Z";
        mUsecaseHandler.execute(fetchJournalEntriesUseCase,
                new FetchJournalEntries.RequestValues(accountIdentifier, dateRange),
                new UseCase.UseCaseCallback<FetchJournalEntries.ResponseValue>() {

                    @Override
                    public void onSuccess(FetchJournalEntries.ResponseValue response) {
                        transactions = response.getJournalEntryList();
                        delegate.onTransactionsFetchCompleted(transactions);
                    }

                    @Override
                    public void onError(String message) {
                        transactions = null;
                    }
                });
    }
}

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/history/TransactionsHistory.kt
package org.mifos.mobilewallet.mifospay.history

import org.mifos.mobilewallet.core.base.TaskLooper
import org.mifos.mobilewallet.core.base.UseCase.UseCaseCallback
import org.mifos.mobilewallet.core.base.UseCaseFactory
import org.mifos.mobilewallet.core.base.UseCaseHandler
import com.mifos.mobilewallet.model.domain.Transaction
import org.mifos.mobilewallet.core.domain.usecase.account.FetchAccount
import org.mifos.mobilewallet.core.domain.usecase.account.FetchAccountTransactions
import org.mifos.mobilewallet.mifospay.history.HistoryContract.TransactionsHistoryAsync
import javax.inject.Inject

class TransactionsHistory @Inject constructor(private val mUsecaseHandler: UseCaseHandler) {
    var delegate: TransactionsHistoryAsync? = null

    @JvmField
    @Inject
    var mFetchAccountUseCase: FetchAccount? = null

    @JvmField
    @Inject
    var fetchAccountTransactionsUseCase: FetchAccountTransactions? = null

    @JvmField
    @Inject
    var mTaskLooper: TaskLooper? = null

    @JvmField
    @Inject
    var mUseCaseFactory: UseCaseFactory? = null
    private var transactions: List<Transaction>?

    init {
        transactions = ArrayList()
    }

    fun fetchTransactionsHistory(accountId: Long) {
        mUsecaseHandler.execute(fetchAccountTransactionsUseCase,
            FetchAccountTransactions.RequestValues(accountId),
            object : UseCaseCallback<FetchAccountTransactions.ResponseValue?> {
                override fun onSuccess(response: FetchAccountTransactions.ResponseValue?) {
                    transactions = response?.transactions
                    delegate!!.onTransactionsFetchCompleted(transactions)
                }

                override fun onError(message: String) {
                    transactions = null
                }
            })
    }
}